crates/solver/src/settlement_ranker.rs

```diff
@@ -191,17 +184,85 @@ impl SettlementRanker {
                 )),
             );
         }
-        for (i, (solver, _, _)) in enumerate(&rated_settlements) {
-            let rank = rated_settlements.len() - i;
-            solver.notify_auction_result(auction_id, AuctionResult::Ranked(rank));
-            self.metrics
-                .settlement_simulation(solver.name(), SolverSimulationOutcome::Success);
-        }
 
+        if cfg!(feature = "CIP17") {
+            // Filter out settlements that have negative score
+            rated_settlements.retain(|(solver, settlement, _)| {
+                if settlement.score < BigRational::zero() {
+                    tracing::debug!(
+                        solver_name = %solver.name(),
+                        "settlement(s) filtered for having negative score",
+                    );
+                    solver.notify_auction_result(
+                        auction_id,
+                        AuctionResult::Rejected(SolverRejectionReason::NegativeScore(
+                            settlement.score.clone(),
+                        )),
+                    );
+                    return false;
+                }
+                true
+            });
+
+            // Before sorting, make sure to shuffle the settlements. This is to make sure we don't give
+            // preference to any specific solver when there is an objective value tie.
+            rated_settlements.shuffle(&mut rand::thread_rng());
```

Comment:
I think we can safely move the `suffle` before the `if cfg!(feature = "CIP17") {} else {}` block, no?